src/traceme.c

```diff
@@ -0,0 +1,431 @@
+#include <sys/ioctl.h>
+#include <semaphore.h>
+
+#define TRACE_OUTPUT    "trace.data"
+#define SYSFS_PT_TYPE   "/sys/bus/event_source/devices/intel_pt/type"
+#define MAX_PT_TYPE_STR 8
+
+#ifndef INFTIM
+#define INFTIM -1
+#endif
+
+/* Debug helpers */
+#define DEBUG(x...)                       \
+    do {                                  \
+        fprintf(stderr, "%s:%d [%s]: ",   \
+           __FILE__, __LINE__, __func__); \
+        fprintf(stderr, x);               \
+        fprintf(stderr, "\n");            \
+    } while (0)
+
+/*
+ * Stores all information about the tracer.
+ * Exposed to Rust only as an opaque pointer.
+ */
+struct tracer_ctx {
+    pthread_t           tracer_thread;      /* Tracer thread handle */
+    sem_t               tracer_init_sem;    /* Tracer init sync */
+    int                 loop_done_fds[2];   /* Tells the poll loop to exit */
+    int                 perf_fd;            /* FD used to talk to the perf API */
+    pid_t               target_pid;         /* PID to trace */
+    int                 out_fd;             /* Trace written here */
+    struct perf_event_mmap_page
+                        *mmap_header;       /* Aux buffer meta-data */
+    void                *aux;               /* Aux buffer itself */
+    size_t              data_bufsize;       /* Data buf size  (in pages) */
+    size_t              aux_bufsize;        /* Aux buf size  (in pages) */
+};
+
+/*
+ * Passed from Rust to C to configure tracing.
+ * Must stay in sync with the Rust-side.
+ */
+struct tracer_conf {
+    pid_t       target_pid;         /* PID to trace */
+    const char  *trace_filename;    /* Filename to store trace into */
+    const char  *map_filename;      /* Filename to copy linker map to */
+    size_t      data_bufsize;       /* Data buf size (in pages) */
+    size_t      aux_bufsize;        /* Aux buf size (in pages) */
+};
+
+/* Private prototypes */
+static void stash_maps(pid_t, const char *);
+static void write_buf_to_disk(int, void *, __u64);
+static void read_circular_buf(void *, __u64, __u64, __u64 *, int);
+static void poll_loop(struct tracer_ctx *);
+static void *tracer_thread(void *);
+
+/* Exposed Prototypes */
+struct tracer_ctx *traceme_start_tracer(struct tracer_conf *);
+void traceme_stop_tracer(struct tracer_ctx *tr_ctx);
+
+/*
+ * Save linker relocation decisions so that you can later recover the
+ * instruction stream from an on-disk binary.
+ */
+static void
+stash_maps(pid_t pid, const char *map_filename)
+{
+    char    *cmd;
+    int     res;
+
+    DEBUG("saving map to %s", map_filename);
+
+    res = asprintf(&cmd, "cp /proc/%d/maps %s", pid, map_filename);
+    if (res == -1) {
+        err(EXIT_FAILURE, "asprintf");
+    }
+
+    res = system(cmd);
+    if (res != 0) {
+        err(EXIT_FAILURE, "system");
+    }
+
+    free(cmd);
+}
+
+/*
+ * Write part of a buffer to a file descriptor.
+ */
+static void
+write_buf_to_disk(int fd, void *buf, __u64 size) {
+    __u64 written = 0;
```

Comment:
I don't think this type is correct: I think it should be `ssize_t`.

Reply:
Because `write(2)` returns a `ssize_t`?

In general I agree, but in this case it's OK as we check the only case where we can have a negative result. By using a 64-bit unsigned we can count higher without an overflow (or in the case of `ssize_t`, undefined behaviour).

Reply:
There's no guarantee that `ssize_t` fits in a `u64` AFAIK.

Reply:
I'm not sure now, but given:

```
size_t is guaranteed to be at least 16 bits wide. Additionally, POSIX includes ssize_t, which is a signed integral type of the same width as size_t.
```
I think we can use `ssize_t` unless we find overflow problems later.

Reply:
1024 bits is bigger than 16 bits, but won't fit in a `u64`...